src/core/ddsi/src/q_entity.c

```diff
@@ -4000,8 +4456,46 @@ void new_proxy_participant
      new_proxy_{writer,reader} to work */
   ephash_insert_proxy_participant_guid (gv->guid_hash, proxypp);
 
-  /* TODO: Do security checks on the proxy participant. Either add the endpoints or delete the proxy. */
+#ifdef DDSI_INCLUDE_SECURITY
+  if (secure)
+  {
+    /* Secure participant detected: start handshake. */
+    if ((plist->present & PP_IDENTITY_TOKEN))
+    {
+      /* initialize the security attributes associated with the proxy participant */
+      q_omg_security_init_remote_participant(proxypp);
+      /* Check if this secure participant is allowed. */
+      if (proxy_participant_check_security_info(gv, proxypp)) {
```

Comment:
I have to think through the implications of creating the proxy participant and registering it in the GUID hash table before checking whether the security info is acceptable. (If the local participant checks against all proxies, and all proxies check against all locals, it is important to do it in this order in order to check all pairs, but making it visible possibly has other consequences as well. Possibly the parallelism needs to be eliminated.)